falcon/asgi/request.py

```diff
@@ -382,18 +397,21 @@ def __init__(self, scope, receive, first_event=None, options=None):
         #  Misc.
         # =====================================================================
 
-        self._asgi_server_cached = None  # Lazy
+        # PERF(vytas): Fall back to class variable(s) when unset.
+        # self._asgi_server_cached = None  # Lazy
```

Comment:
Since have `__dict__`  can't we just use [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property)?
It's easy to backport, but I'm have not done benchmarks to see if it's faster than out current solution.

an alternative is to remove `__dict__`

Reply:
`cached_property` is an interesting idea, but it is 3.8+ only... Worth benchmarking tho anyway.

Reply:
> 
> 
> `cached_property` is an interesting idea, but it is 3.8+ only... Worth benchmarking tho anyway.
> Assuming it is implemented in C like that nifty `lru_cache`, otherwise it won't be much better than our stuff.

No, it's a pure python implementation and it can be backported on previous versions without problems (using it on 3.7 at work without issues).
Regarding performance vs a classic `@property`, it should be faster after the first access, since it's just a normal attribute access. I'm not sure regarding the first access though.
Given out use case, I think that most properties are used a low number of times (and I would hazard only once in a fair number of cases) so first access performance is fairly important.

Reply:
We can take a look at faster ways to cache properties in 3.1.

However, if (as you point out) this `cached_property` is a generic recipe implemented in pure Python, it is unlikely to provide any improvement over our implementation.